### ISSUE TITLE
[net6] Extensions crash linker during build

### ISSUE THREAD

**Issue by chamons** [xamarin/xamarin-macios#12692]
### Steps to Reproduce

1. Build a net6 extension project. For internal folks we can use the submission test:
    make SUBMISSION_FILTERS="--filter 'ODRsTVOS_Extension (.NET)'"
2. Crash in build

### Expected Behavior

Build to complete. 

### Actual Behavior

```
/Users/donblas/Programming/macios/master/xamarin-macios/builds/downloads/dotnet-sdk-6.0.100-rc.2.21456.8-osx-x64/packs/Microsoft.tvOS.Sdk/15.0.100-ci.xtro-fixes.564/targets/Xamarin.Shared.Sdk.targets(799,3): warning : Some features may not work correctly, because the generation of the runtime configure file (*.runtimeconfig.json) has been disabled. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
ILLink : TVOS error IL6999: Failed to execute the custom steps. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
ILLINK : error MT2362: The linker step 'Registrar' failed during processing: Could not register the assembly 'Xamarin.TVOS': System.NotImplementedException: The method or operation is not implemented. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
     at Xamarin.Linker.DotNetResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters) in /Users/donblas/Programming/macios/master/xamarin-macios/tools/dotnet-linker/DotNetResolver.cs:line 11
     at Registrar.StaticRegistrar.GetSystemVoidType() in /Users/donblas/Programming/macios/master/xamarin-macios/tools/common/StaticRegistrar.cs:line 996
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 2166
     at Registrar.Registrar.GetProtocols(ObjCType type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1726
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1986
     at Registrar.Registrar.GetProtocols(ObjCType type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1726
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1986
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1964
     at Registrar.Registrar.RegisterAssembly(AssemblyDefinition assembly) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 2486
  		
  	Could not register the assembly 'Xamarin.TVOS': System.NotImplementedException: The method or operation is not implemented.
     at Xamarin.Linker.DotNetResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters) in /Users/donblas/Programming/macios/master/xamarin-macios/tools/dotnet-linker/DotNetResolver.cs:line 11
     at Registrar.StaticRegistrar.GetSystemVoidType() in /Users/donblas/Programming/macios/master/xamarin-macios/tools/common/StaticRegistrar.cs:line 996
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 2166
     at Registrar.Registrar.GetProtocols(ObjCType type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1726
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1986
     at Registrar.Registrar.GetProtocols(ObjCType type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1726
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1986
     at Registrar.Registrar.RegisterTypeUnsafe(TypeReference type, List`1& exceptions) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 1964
     at Registrar.Registrar.RegisterAssembly(AssemblyDefinition assembly) in /Users/donblas/Programming/macios/master/xamarin-macios/src/ObjCRuntime/Registrar.cs:line 2486
  		
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
/Users/donblas/Programming/macios/master/xamarin-macios/builds/downloads/dotnet-sdk-6.0.100-rc.2.21456.8-osx-x64/sdk/6.0.100-rc.2.21456.8/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ILLink.targets(108,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]

Build FAILED.

/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOS/obj/iPhone/Release/net6.0-tvos/tvos-arm64/ODRsTVOS/Assets.xcassets : actool warning : The data set "TestDataAsset" has 14 unassigned children. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOS/ODRsTVOS.csproj]
/Users/donblas/Programming/macios/master/xamarin-macios/builds/downloads/dotnet-sdk-6.0.100-rc.2.21456.8-osx-x64/packs/Microsoft.tvOS.Sdk/15.0.100-ci.xtro-fixes.564/targets/Xamarin.Shared.Sdk.targets(799,3): warning : Some features may not work correctly, because the generation of the runtime configure file (*.runtimeconfig.json) has been disabled. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
ILLink : TVOS error IL6999: Failed to execute the custom steps. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
ILLINK : error MT2362: The linker step 'Registrar' failed during processing: Could not register the assembly 'Xamarin.TVOS': System.NotImplementedException: The method or operation is not implemented. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
/Users/donblas/Programming/macios/master/xamarin-macios/builds/downloads/dotnet-sdk-6.0.100-rc.2.21456.8-osx-x64/sdk/6.0.100-rc.2.21456.8/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ILLink.targets(108,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [/Users/donblas/Programming/macios/master/SubmissionSamples/ODRsTVOS_Extension/dotnet/ODRsTVOSExt/ODRsTVOSExt.csproj]
    2 Warning(s)
    3 Error(s)

Time Elapsed 00:00:22.93
```

### Environment

Latest main

**Comment by chamons** [comment]
So I can reproduce this 100% of the time inside the debugger, and here's what I'm seeing:

- We are crashing [here](https://github.com/xamarin/xamarin-macios/blob/main/tools/common/StaticRegistrar.cs#L996) trying to resolve System.Private.CoreLib.
- Which is not in our list of input_assemblies
- We are doing this while trying to process `encodeWithCoder:` on 	`Foundation.INSCoding` which is a protocol on 	UNNotificationTrigger.
- From what I can tell UNNotificationTrigger is not being used directly.
- Adding a usage of it in a working iOS app does not change behavior.

**Comment by chamons** [comment]
In discussions with Rolf, we've discovered that System.Private.CoreLib.dll is not being referenced because we're only getting reference assemblies not implementation ones.

Going to dig for where that swap happens.

**Comment by rolfbjarne** [comment]
This was fixed in #12703.
